src/script.c

```diff
@@ -1009,3 +1009,117 @@ int wally_elements_pegin_contract_script_from_bytes(const unsigned char *redeem_
+{
+    unsigned char *p = bytes, *end = p + bytes_len;
+    struct wally_tx_witness_stack *result = NULL;
+    int ret = WALLY_OK;
+
+    if (!bytes || !output || !bytes_len) {
+        return WALLY_EINVAL;
+    }
+
+    if ((ret = wally_tx_witness_stack_init_alloc(2, &result)) != WALLY_OK) {
+        return ret;
+    }
+
+    while (p < end) {
+        size_t push_size, push_opcode_size;
+
+        if ((ret = script_get_push_size_from_bytes(p, end - p, &push_size)) != WALLY_OK) {
+            goto fail;
+        }
+        if ((ret = script_get_push_opcode_size_from_bytes(p, end - p, &push_opcode_size)) != WALLY_OK) {
+            goto fail;
+        }
+        p += push_opcode_size;
+
+        if ((ret = wally_tx_witness_stack_add(result, p, push_size)) != WALLY_OK) {
+            goto fail;
+        }
+        p += push_size;
+    }
+
+    *output = result;
+    return WALLY_OK;
+
+fail:
+    wally_tx_witness_stack_free(result);
+    return ret;
+}
+
+int wally_witness_p2wpkh_from_der(
+    const unsigned char *pub_key,
+    size_t pub_key_len,
+    const unsigned char *sig,
+    size_t sig_len,
+    struct wally_tx_witness_stack **witness)
+{
+    unsigned char script_sig[WALLY_SCRIPTSIG_P2PKH_MAX_LEN];
+    int ret;
+    size_t script_sig_len;
+
+    if ((ret = wally_scriptsig_p2pkh_from_der(pub_key, pub_key_len, sig, sig_len, script_sig, WALLY_SCRIPTSIG_P2PKH_MAX_LEN, &script_sig_len)) != WALLY_OK) {
+        return ret;
+    }
+
+    if ((ret = scriptsig_to_witness(script_sig, script_sig_len, witness)) != WALLY_OK) {
+        return ret;
+    }
+
+    return WALLY_OK;
+}
+
+int wally_witness_p2wpkh_from_sig(
+    const unsigned char *pub_key,
+    size_t pub_key_len,
+    const unsigned char *sig,
+    size_t sig_len,
+    uint32_t sighash,
+    struct wally_tx_witness_stack **witness)
+{
+    unsigned char script_sig[WALLY_SCRIPTSIG_P2PKH_MAX_LEN];
+    int ret;
+    size_t script_sig_len;
+
+    if ((ret = wally_scriptsig_p2pkh_from_sig(pub_key, pub_key_len, sig, sig_len, sighash, script_sig, WALLY_SCRIPTSIG_P2PKH_MAX_LEN, &script_sig_len)) != WALLY_OK) {
+        return ret;
+    }
+
+    if ((ret = scriptsig_to_witness(script_sig, script_sig_len, witness)) != WALLY_OK) {
+        return ret;
+    }
+
+    return WALLY_OK;
+}
+
+int wally_witness_multisig_from_bytes(
+    const unsigned char *script,
+    size_t script_len,
+    const unsigned char *bytes,
+    size_t bytes_len,
+    const uint32_t *sighash,
+    size_t sighash_len,
+    uint32_t flags,
+    struct wally_tx_witness_stack **witness)
+{
+    unsigned char *script_sig = NULL;
+    int ret = WALLY_OK;
+    size_t script_sig_len, n_sigs, buf_len;
+
+    if (!script || !script_len || !bytes || !bytes_len || !sighash || !sighash_len || !witness || !script_is_op_n(script[0], false, &n_sigs)) {
+        return WALLY_EINVAL;
+    }
+
+    buf_len = n_sigs * (EC_SIGNATURE_DER_MAX_LEN + 2) + script_len;
+    script_sig = wally_malloc(buf_len);
```

Comment:
Maybe explicitly check for null and return WALLY_ENOMEM?

Reply:
Done